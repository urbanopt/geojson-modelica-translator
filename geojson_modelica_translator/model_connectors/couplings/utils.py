# :copyright (c) URBANopt, Alliance for Sustainable Energy, LLC, and other contributors.
# See also https://github.com/urbanopt/geojson-modelica-translator/blob/develop/LICENSE.md

import re
from collections import deque, namedtuple

DiagramTransformation = namedtuple('DiagramTransformation', ['model_name', 'model_type'])
DiagramLine = namedtuple('DiagramLine', ['a_name', 'a_port', 'b_name', 'b_port'])

JINJA_EXPRESSION_REGEX = re.compile(r'{{\s*(.*?)\s*}}')


def parse_diagram_command(str_cmd):
    """Returns a diagram command or None if it's not a diagram command"""
    cmd_args_list = str_cmd.split('.')
    if len(cmd_args_list) <= 1:
        return None

    context = cmd_args_list.pop(0)
    if context != 'diagram':
        return None

    command = cmd_args_list.pop(0)
    if command == 'transformation':
        if len(cmd_args_list) != 2:
            raise Exception(f'Invalid diagram templating command: "transformation" expects 2 arguments but got {cmd_args_list}')
        return DiagramTransformation(*cmd_args_list)
    elif command == 'line':
        if len(cmd_args_list) != 4:
            raise Exception(f'Invalid diagram templating command: "line" expects 4 arguments but got {cmd_args_list}')
        return DiagramLine(*cmd_args_list)
    else:
        raise Exception(f'Invalid diagram templating command "{command}"')


def parse_diagram_commands(template_contents):
    """Returns a list of diagram commands parsed from the template modelica contents
    i.e. it will find any instances of {{ diagram.<cmd>.* }} in the template

    :param template_contents: str, modelica template code
    :return: list[DiagramCommand]
    """
    matches = JINJA_EXPRESSION_REGEX.finditer(template_contents)

    commands = []
    for match in matches:
        group = match.group(1)
        diagram_command = parse_diagram_command(group)
        if diagram_command is None:
            continue

        commands.append(diagram_command)

    return commands


def find_path_bfs(matrix, start_row, start_col, end_row, end_col):
    """Find path from start to end in the matrix. Matrix items that are None are
    considered "empty" and traversable. Raises an exception if no path is found.

    :param matrix: list[list[]]
    :param start_row: int
    :param start_col: int
    :param end_row: int
    :param end_col: int
    :return: list, list of (row, col) tuples
    """
    start = (start_row, start_col)
    finish = (end_row, end_col)

    queue = deque([([], start)])
    visited = []

    def visitable(pos):
        return (
            # must be valid matrix position
            pos[0] >= 0
            and pos[0] < len(matrix)
            and pos[1] >= 0
            and pos[1] < len(matrix[0])
            # must be empty or the finish
            and (matrix[pos[0]][pos[1]] is None or pos == finish)
            # must not be visited
            and pos not in visited
        )

    def get_neighbors(pos):
        north = (pos[0] - 1, pos[1])
        south = (pos[0] + 1, pos[1])
        east = (pos[0], pos[1] - 1)
        west = (pos[0], pos[1] + 1)
        neighbors = [north, south, east, west]
        return [direction for direction in neighbors if visitable(direction)]

    while queue:
        path, current = queue.popleft()

        if current == finish:
            return path
        if current in visited:
            continue

        visited.append(current)

        for neighbor in get_neighbors(current):
            queue.append((path + [neighbor], neighbor))

    # failed to find a path
    raise Exception(f'Failed to find path from {start} to {finish}')
