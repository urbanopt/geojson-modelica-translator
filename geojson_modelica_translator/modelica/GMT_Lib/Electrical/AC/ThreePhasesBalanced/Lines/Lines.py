from pathlib import Path

from geojson_modelica_translator.modelica.simple_gmt_base import SimpleGMTBase


class DistributionLines(SimpleGMTBase):
    def __init__(self, system_parameters):
        self.system_parameters = system_parameters
        self.template_dir = Path(__file__).parent
        super().__init__(self.system_parameters, self.template_dir)

    def build_from_template(self, output_dir: Path):
        distribution_line_params = self.system_parameters.get_param("$.distribution_lines")
        # Each segment of cable gets its own mofile, numbered with 'iteration'
        for index, line in enumerate(distribution_line_params):
            # This is how we map from ditto-reader to mbl. It's pretty brittle, and will need updating in this state
            # https://github.com/urbanopt/urbanopt-ditto-reader/blob/develop/example/extended_catalog.json
            for wire in line["commercial_line_type"]:
                if '477kcmil' in wire:
                    mbl_wire = 'Buildings.Electrical.Transmission.MediumVoltageCables.Annealed_Al_500'
                if '750kcmil' in wire:
                    # FIXME: This mapping from 750kcmil to 1000kcmil is not ideal
                    # A 750kcmil has been proposed for the MBL
                    # The temporary alternative is to map to 500 kcmil, which didn't seem right to me
                    mbl_wire = 'Buildings.Electrical.Transmission.MediumVoltageCables.Annealed_Al_1000'
            line_params = {
                'length': line["length"],
                'ampacity': line["ampacity"],
                'nominal_voltage': line["nominal_voltage"],
                'commercial_line_type': mbl_wire,
                'model_name': f"Line{index}",
            }
            # render template to final modelica file
            self.to_modelica(output_dir=output_dir, model_name='ACLine', param_data=line_params, iteration=index)
