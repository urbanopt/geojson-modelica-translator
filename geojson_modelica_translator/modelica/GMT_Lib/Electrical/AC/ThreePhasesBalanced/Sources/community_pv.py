from pathlib import Path

from geojson_modelica_translator.modelica.simple_gmt_base import SimpleGMTBase


class CommunityPV(SimpleGMTBase):
    def __init__(self, system_parameters):
        self.system_parameters = system_parameters
        self.template_dir = Path(__file__).parent
        super().__init__(self.system_parameters, self.template_dir)

    def build_from_template(self, output_dir: Path):
        community_pv_params = self.system_parameters.get_param("$.photovoltaic_panels")
        # There can be multiple community pv arrays so we need to loop over them
        for index, pvarray in enumerate(community_pv_params):
            pv_params = {
                'net_surface_area_m2': pvarray["net_surface_area"],
                'nominal_voltage_V': pvarray["nominal_voltage"],
                'surface_azimuth_deg': pvarray["surface_azimuth"],
                'surface_tilt_deg': pvarray["surface_tilt"],
                'model_name': f"PVPanels{index}",
            }
            # render template to final modelica file
            self.to_modelica(output_dir=output_dir, model_name='PVPanels', param_data=pv_params, iteration=index)
