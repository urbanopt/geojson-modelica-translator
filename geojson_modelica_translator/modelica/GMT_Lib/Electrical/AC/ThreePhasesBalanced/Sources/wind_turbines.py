from pathlib import Path

from geojson_modelica_translator.modelica.simple_gmt_base import SimpleGMTBase


class WindTurbine(SimpleGMTBase):
    def __init__(self, system_parameters):
        self.system_parameters = system_parameters
        self.template_dir = Path(__file__).parent
        super().__init__(self.system_parameters, self.template_dir)

    def build_from_template(self, output_dir: Path):
        wind_turbine_params = self.system_parameters.get_param("$.wind_turbines")
        # There can be multiple wind turbines so we need to loop over them
        for index, turbine in enumerate(wind_turbine_params):
            turbine_params = {
                'scaling_factor': turbine["scaling_factor"],
                'height_over_ground': turbine["height_over_ground"],
                'nominal_voltage': turbine["nominal_voltage"],
                'power_curve': turbine["power_curve"],
                'model_name': f"WindTurbine{index}",
            }
            # FIXME: "power_curve" is a list of lists in the sys-param file.
            # This whole string shenanigans is to format it into something that will make modelica happy.
            # TODO: confirm that modelica chokes on a list of lists - ie comment this whole block out and see if the model will run.
            power_curve_string = '['
            for point in turbine_params["power_curve"]:
                power_curve_string += f'{point[0]}, {point[1]}; '
            power_curve_string = power_curve_string.rstrip('; ')
            power_curve_string += ']'
            turbine_params["power_curve"] = power_curve_string
            # render template to final modelica file
            self.to_modelica(output_dir=output_dir, model_name='WindTurbine', param_data=turbine_params, iteration=index)
