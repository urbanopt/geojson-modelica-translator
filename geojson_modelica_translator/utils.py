# :copyright (c) URBANopt, Alliance for Sustainable Energy, LLC, and other contributors.
# See also https://github.com/urbanopt/geojson-modelica-translator/blob/develop/LICENSE.md

import logging
import os
import shutil
from pathlib import Path
from uuid import uuid4

_log = logging.getLogger(__name__)


def copytree(src, dst, symlinks=False, ignore=None):
    """
    Alternate version of copytree that will work if the directory already exists (use instead of shutil)
    """
    for item in os.listdir(src):
        s = os.path.join(src, item)
        d = os.path.join(dst, item)
        if os.path.isdir(s):
            shutil.copytree(s, d, symlinks, ignore)
        else:
            shutil.copy2(s, d)


def convert_c_to_k(c):
    """Converts a temperature in celsius to kelvin

    :param c: float, temperature in celsius
    :return: float, temperature in kelvin
    """
    return c + 273.15


def linecount(filename: Path) -> int:
    """Counts the number of lines in a file
    Probably not the most efficient way to do this, but it works
    """
    return len(open(filename).readlines())


class ModelicaPath(object):
    """
    Class for storing Modelica paths. This allows the path to point to
    the model directory, resources, and scripts directory.
    """

    def __init__(self, name, root_dir, overwrite=False):
        """
        Create a new modelica-based path with name of 'name'

        :param name: Name to create
        """
        self.name = name
        self.root_dir = root_dir
        self.overwrite = overwrite

        # create the directories
        if root_dir is not None:
            check_path = os.path.join(self.files_dir)
            self.clear_or_create_path(check_path)
            check_path = os.path.join(self.resources_dir)
            self.clear_or_create_path(check_path)
            check_path = os.path.join(self.scripts_dir)
            self.clear_or_create_path(check_path)

    def clear_or_create_path(self, path):
        if os.path.exists(path):
            if not self.overwrite:
                raise Exception("Directory already exists and overwrite is false for %s" % path)
            else:
                shutil.rmtree(path)
        os.makedirs(path, exist_ok=True)

    @property
    def files_dir(self):
        """
        Return the path to the files (models) for the specified ModelicaPath. This path does not include the
        trailing slash.

        :return: string, path to where files (models) are stored, without trailing slash
        """
        if self.root_dir is None:
            return self.files_relative_dir
        else:
            return f"{self.root_dir}/{self.name}"

    @property
    def resources_relative_dir(self):
        """
        Return the relative resource directory instead of the full path. This is useful when replacing
        strings within modelica files which are relative to the package.

        :return: string, relative resource's data path
        """
        return f"Resources/Data/{self.name}"

    @property
    def scripts_relative_dir(self, platform='Dymola'):
        """Return the scripts directory that is in the resources directory. This only returns the
        relative directory and is useful when replacing string values within Modelica files.

        :return: string, relative scripts path
        """
        return f"Resources/Scripts/{self.name}/{platform}"

    @property
    def files_relative_dir(self):
        """Return the path to the files relative to the project name."""
        return os.path.join(self.name)

    @property
    def resources_dir(self):
        """
        Return the path to the resources directory for the specified ModelicaPath. This path does not include
        the trailing slash.

        :return: string, path to where resources are stored, without trailing slash.
        """
        if self.root_dir is None:
            return self.resources_relative_dir
        else:
            return f"{self.root_dir}/{self.resources_relative_dir}"

    @property
    def scripts_dir(self):
        """
        Return the path to the scripts directory (in the resources dir) for the specified ModelicaPath.
        This path does not include the trailing slash.

        :return: string, path to where scripts are stored, without trailing slash.
        """
        if self.root_dir is None:
            return self.scripts_relative_dir
        else:
            return f"{self.root_dir}/{self.scripts_relative_dir}"


# This is used for some test cases where we need deterministic IDs to be generated
USE_DETERMINISTIC_ID = bool(os.environ.get('GMT_DETERMINISTIC_ID', False))

counter = 0


def simple_uuid():
    """Generates a simple string uuid

    :return: string, uuid
    """
    global counter

    if not USE_DETERMINISTIC_ID:
        return str(uuid4()).split("-")[0]
    else:
        id = str(counter)
        counter += 1
        return id
