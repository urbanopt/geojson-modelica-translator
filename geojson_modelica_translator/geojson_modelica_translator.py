# :copyright (c) URBANopt, Alliance for Sustainable Energy, LLC, and other contributors.
# See also https://github.com/urbanopt/geojson-modelica-translator/blob/develop/LICENSE.md

import logging
from pathlib import Path

from geojson_modelica_translator.geojson.urbanopt_geojson import (
    UrbanOptGeoJson
)
from geojson_modelica_translator.model_connectors.couplings import (
    Coupling,
    CouplingGraph
)
from geojson_modelica_translator.model_connectors.districts import District
from geojson_modelica_translator.model_connectors.energy_transfer_systems import (
    CoolingIndirect,
    HeatingIndirect
)
from geojson_modelica_translator.model_connectors.load_connectors import (
    Spawn,
    Teaser,
    TimeSeries,
    TimeSeriesMFT
)
from geojson_modelica_translator.model_connectors.networks import Network2Pipe
from geojson_modelica_translator.model_connectors.networks.network_distribution_pump import (
    NetworkDistributionPump
)
from geojson_modelica_translator.model_connectors.plants import CoolingPlant
from geojson_modelica_translator.model_connectors.plants.borefield import (
    Borefield
)
from geojson_modelica_translator.model_connectors.plants.chp import (
    HeatingPlantWithOptionalCHP
)
from geojson_modelica_translator.modelica.modelica_runner import ModelicaRunner
from geojson_modelica_translator.system_parameters.system_parameters import (
    SystemParameters
)

_log = logging.getLogger(__name__)


# map the system parameter "load_model" to Python class
LOAD_MODEL_TO_CLASS = {
    "spawn": Spawn,
    "rc": Teaser,
    "time_series": TimeSeries,
    "time_series_massflow_temperature": TimeSeriesMFT,
}


def _parse_couplings(geojson, sys_params, district_type=None):
    """Given config files, construct the necessary models and their couplings which
    can then be passed to CouplingGraph.

    :param geojson: UrbanOptGeoJson
    :param sys_params: SystemParameters
    :param district_type: str - type of district [None, "4G", "5G"]
    :return: list[Coupling], list of couplings to be passed to CouplingGraph
    """
    # 4G implementation assumes that all generated district energy system models will have:
    #   - one heating plant
    #   - one cooling plant
    #   - one heating distribution network
    #   - one cooling distribution network
    #   - one heating and cooling ETS per load
    # NOTE: loads can be of any type/combination
    all_couplings = []

    if district_type is None or district_type == "4G":
        # create the plants and networks
        cooling_network = Network2Pipe(sys_params)
        cooling_plant = CoolingPlant(sys_params)
        heating_network = Network2Pipe(sys_params)
        heating_plant = HeatingPlantWithOptionalCHP(sys_params)
        all_couplings += [
            Coupling(cooling_plant, cooling_network),
            Coupling(heating_plant, heating_network),
        ]
    elif district_type == "5G":
        # create ambient water stub
        ambient_water_stub = NetworkDistributionPump(sys_params)
        # create borefield
        borefield = Borefield(sys_params)
        all_couplings.append(Coupling(borefield, ambient_water_stub, district_type))
        all_couplings.append(Coupling(ambient_water_stub, ambient_water_stub, district_type))

    # create the loads and their ETSes
    for building in geojson.buildings:
        load_model_type = sys_params.get_param_by_building_id(building.id, "load_model")
        load_class = LOAD_MODEL_TO_CLASS[load_model_type]
        load = load_class(sys_params, building)

        if district_type is None or district_type == "4G":
            cooling_indirect = CoolingIndirect(sys_params, building.id)
            all_couplings.append(Coupling(load, cooling_indirect))
            all_couplings.append(Coupling(cooling_indirect, cooling_network))

            heating_indirect = HeatingIndirect(sys_params, building.id)
            all_couplings.append(Coupling(load, heating_indirect))
            all_couplings.append(Coupling(heating_indirect, heating_network))
        elif district_type == "5G":
            all_couplings.append(Coupling(load, ambient_water_stub, district_type))

    return all_couplings


class ModelicaPackage(object):
    """Represents a modelica package which can be simulated"""

    def __init__(self, file_to_run, project_path, project_name):
        self._file_to_run = file_to_run
        self._project_path = project_path
        self._project_name = project_name

    def simulate(self):
        """Simulate the package.

        :return: tuple(bool, pathlib.Path), True or False depending on simulation success
            followed by the path to the results directory
        """
        modelica_runner = ModelicaRunner()
        return modelica_runner.run_in_docker(
            self._file_to_run,
            run_path=self._project_path,
            project_name=self._project_name
        )


class GeoJsonModelicaTranslator(object):
    """
    Main class for using the GeoJSON to Modelica Translator.
    """

    def __init__(
        self,
        geojson_filepath,
        sys_params_filepath,
        root_dir,
        project_name,
    ):
        """Create an instance of this class

        :param geojson_filepath: str, path to GeoJSON file
        :param sys_params_filepath: str, path to system parameters file
        :param root_dir: str, where to create the package
        :project_name: str, name of the package
        """
        if not Path(geojson_filepath).exists():
            raise FileNotFoundError(f'GeoJSON file path does not exist: {geojson_filepath}')
        if not Path(sys_params_filepath).exists():
            raise FileNotFoundError(f'System parameters file path does not exist: {sys_params_filepath}')

        self._system_parameters = SystemParameters(sys_params_filepath)

        geojson_ids = self._system_parameters.get_default(
            '$.buildings.[*].geojson_id',
            []
        )
        self._geojson = UrbanOptGeoJson(geojson_filepath, geojson_ids)

        # Use different couplings for each district system type
        district_type = self._system_parameters.get_param("district_system")
        if 'fifth_generation' in district_type:
            self._couplings = _parse_couplings(self._geojson, self._system_parameters, district_type='5G')
        elif 'fourth_generation' in district_type:
            self._couplings = _parse_couplings(self._geojson, self._system_parameters)

        self._root_dir = root_dir
        self._project_name = project_name
        self._coupling_graph = CouplingGraph(self._couplings)
        self._district = District(
            self._root_dir,
            self._project_name,
            self._system_parameters,
            self._coupling_graph
        )
        self._package_created = False

    def to_modelica(self):
        """Generate the modelica package. Call `simulate` method on the result
        to run the package

        :return: ModelicaPackage
        """
        self._district.to_modelica()

        return ModelicaPackage(
            self._district.district_model_filepath,
            self._root_dir,
            self._project_name
        )
