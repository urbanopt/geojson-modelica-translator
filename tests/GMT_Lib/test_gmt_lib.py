"""
****************************************************************************************************
:copyright (c) 2019-2022, Alliance for Sustainable Energy, LLC, and other contributors.

All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted
provided that the following conditions are met:

Redistributions of source code must retain the above copyright notice, this list of conditions
and the following disclaimer.

Redistributions in binary form must reproduce the above copyright notice, this list of conditions
and the following disclaimer in the documentation and/or other materials provided with the
distribution.

Neither the name of the copyright holder nor the names of its contributors may be used to endorse
or promote products derived from this software without specific prior written permission.

Redistribution of this software, without modification, must refer to the software by the same
designation. Redistribution of a modified version of this software (i) may not refer to the
modified version by the same designation, or by any confusingly similar designation, and
(ii) must refer to the underlying software originally provided by Alliance as “URBANopt”. Except
to comply with the foregoing, the term “URBANopt”, or any confusingly similar designation may
not be used to refer to any modified version of this software or any modified version of the
underlying software originally provided by Alliance without the prior written consent of Alliance.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
****************************************************************************************************
"""
from pathlib import Path
from shutil import copyfile

import pytest
from geojson_modelica_translator.modelica.GMT_Lib.Electrical.AC.ThreePhasesBalanced.Lines.Lines import (
    DistributionLines
)
from geojson_modelica_translator.modelica.GMT_Lib.Electrical.AC.ThreePhasesBalanced.Sources.community_pv import (
    CommunityPV
)
from geojson_modelica_translator.modelica.GMT_Lib.Electrical.AC.ThreePhasesBalanced.Sources.wind_turbines import (
    WindTurbine
)
from geojson_modelica_translator.modelica.modelica_runner import ModelicaRunner
from geojson_modelica_translator.system_parameters.system_parameters import (
    SystemParameters
)
from geojson_modelica_translator.utils import linecount
from jinja2 import Environment, FileSystemLoader, StrictUndefined

PARENT_DIR = Path(__file__).parent
GMT_LIB_PATH = PARENT_DIR.parent.parent / 'geojson_modelica_translator' / 'modelica' / 'GMT_Lib'
COOLING_PLANT_PATH = GMT_LIB_PATH / 'DHC' / 'Components' / 'Plants' / 'Cooling'
MICROGRID_PARAMS = PARENT_DIR.parent / 'data_shared' / 'system_params_microgrid_example.json'

env = Environment(
    loader=FileSystemLoader(GMT_LIB_PATH),
    undefined=StrictUndefined,
    variable_start_string='{$',
    variable_end_string='$}'
)

COOLING_PLANT_PARAMS = {
    'chiller_performance': 'Buildings.Fluid.Chillers.Data.ElectricEIR.ElectricEIRChiller_York_YT_1055kW_5_96COP_Vanes',
    'plant_type': 'Buildings.Experimental.DHC.Plants.Cooling.ElectricChillerParallel',
    'delta_temp_approach': 3,
    'chw_mass_flow_nominal': 18.3,
    'chw_pressure_drop_nominal': 44800,
    'chiller_water_flow_minimum': 0.03,
    'cw_mass_flow_nominal': 34.7,
    'cw_pressure_drop_nominal': 46200,
    'fan_power': 4999,
    'chw_temp_setpoint': 281.15,
    'cooling_tower_pressure_drop_valve_nominal': 5999,
    'chw_pressure_drop_valve_nominal': 5999,
    'cw_pressure_drop_valve_nominal': 5999
}


def test_generate_cooling_plant(snapshot):
    # -- Setup
    template_path = (COOLING_PLANT_PATH / 'CoolingPlant.mot').relative_to(GMT_LIB_PATH)

    # -- Act
    actual = env.get_template(template_path.as_posix()).render(**COOLING_PLANT_PARAMS)

    # -- Assert
    assert actual == snapshot


def test_build_community_pv():
    # -- Setup

    package_output_dir = PARENT_DIR / 'output' / 'CommunityPV'
    package_output_dir.mkdir(parents=True, exist_ok=True)
    sys_params = SystemParameters(MICROGRID_PARAMS)

    # -- Act
    cpv = CommunityPV(sys_params)
    cpv.build_from_template(package_output_dir)

    # -- Assert
    # Did the mofile get created?
    assert linecount(package_output_dir / 'PVPanels1.mo') > 20


@pytest.mark.simulation
def test_simulate_community_pv():
    # -- Setup

    package_output_dir = PARENT_DIR / 'output' / 'CommunityPV'
    package_output_dir.mkdir(parents=True, exist_ok=True)
    sys_params = SystemParameters(MICROGRID_PARAMS)

    # -- Act
    cpv = CommunityPV(sys_params)
    cpv.build_from_template(package_output_dir)

    runner = ModelicaRunner()
    success, _ = runner.run_in_docker(package_output_dir / 'PVPanels0.mo')

    # -- Assert
    # Did the mofile get created?
    assert linecount(package_output_dir / 'PVPanels1.mo') > 20
    # Did the simulation run?
    assert success is True


def test_build_cooling_plant():
    # -- Setup
    template_path = (COOLING_PLANT_PATH / 'CoolingPlant.mot').relative_to(GMT_LIB_PATH)

    # -- Act
    output = env.get_template(template_path.as_posix()).render(**COOLING_PLANT_PARAMS)
    package_output_dir = PARENT_DIR / 'output' / 'Cooling'
    package_output_dir.mkdir(parents=True, exist_ok=True)
    with open(package_output_dir / 'CoolingPlant.mo', 'w') as f:
        f.write(output)

    # -- Assert
    # Did the mofile get created?
    assert linecount(package_output_dir / 'CoolingPlant.mo') > 20


@pytest.mark.simulation
def test_simulate_cooling_plant():
    # -- Setup
    template_path = (COOLING_PLANT_PATH / 'CoolingPlant.mot').relative_to(GMT_LIB_PATH)
    output = env.get_template(template_path.as_posix()).render(**COOLING_PLANT_PARAMS)
    package_output_dir = PARENT_DIR / 'output' / 'Cooling'
    package_output_dir.mkdir(parents=True, exist_ok=True)
    with open(package_output_dir / 'CoolingPlant.mo', 'w') as f:
        f.write(output)

    # copy over the script
    copyfile(COOLING_PLANT_PATH / 'CoolingPlant.mos', package_output_dir / 'CoolingPlant.mos')

    # -- Act
    runner = ModelicaRunner()
    success, _ = runner.run_in_docker(package_output_dir / 'CoolingPlant.mos', package_output_dir, 'Cooling')

    # -- Assert
    assert success is True


def test_build_wind_turbine():
    # -- Setup
    package_output_dir = PARENT_DIR / 'output' / 'WindTurbine'
    package_output_dir.mkdir(parents=True, exist_ok=True)
    sys_params = SystemParameters(MICROGRID_PARAMS)

    # -- Act
    cpv = WindTurbine(sys_params)
    cpv.build_from_template(package_output_dir)

    # -- Assert
    # Did the mofile get created?
    assert linecount(package_output_dir / 'WindTurbine0.mo') > 20


@pytest.mark.simulation
def test_simulate_wind_turbine():
    # -- Setup

    package_output_dir = PARENT_DIR / 'output' / 'WindTurbine'
    package_output_dir.mkdir(parents=True, exist_ok=True)
    sys_params = SystemParameters(MICROGRID_PARAMS)

    # -- Act
    cpv = WindTurbine(sys_params)
    cpv.build_from_template(package_output_dir)

    runner = ModelicaRunner()
    success, _ = runner.run_in_docker(package_output_dir / 'WindTurbine0.mo')

    # -- Assert
    # Did the mofile get created?
    assert linecount(package_output_dir / 'WindTurbine0.mo') > 20
    # Did the simulation run?
    assert success is True


def test_build_distribution_lines():
    # -- Setup
    package_output_dir = PARENT_DIR / 'output' / 'DistributionLines'
    package_output_dir.mkdir(parents=True, exist_ok=True)
    sys_params = SystemParameters(MICROGRID_PARAMS)

    # -- Act
    cpv = DistributionLines(sys_params)
    cpv.build_from_template(package_output_dir)

    # -- Assert
    # Did the mofile get created?
    assert linecount(package_output_dir / 'ACLine0.mo') > 20


@pytest.mark.simulation
def test_simulate_distribution_lines():
    # -- Setup
    package_output_dir = PARENT_DIR / 'output' / 'DistributionLines'
    package_output_dir.mkdir(parents=True, exist_ok=True)
    sys_params = SystemParameters(MICROGRID_PARAMS)

    # -- Act
    cpv = DistributionLines(sys_params)
    cpv.build_from_template(package_output_dir)

    runner = ModelicaRunner()
    success, _ = runner.run_in_docker(package_output_dir / 'ACLine0.mo')

    # -- Assert
    # Did the mofile get created?
    assert linecount(package_output_dir / 'ACLine0.mo') > 20
    # Did the simulation run?
    assert success is True


# Keeping the code below because it may come back and this was a weird issue.
# @pytest.mark.simulation
# def test_stub_mbl_v9_with_not_msl_v4():
#     """Need to have a stub where mbl_v9 is selected that is simulatable (with
#     MSV V3.2) in order to not create a failed pytest command with exit code 5."""
#     assert False is not True
