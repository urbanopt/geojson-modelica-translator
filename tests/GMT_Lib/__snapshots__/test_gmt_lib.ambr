# name: test_generate_cooling_plant
  '''
  within GMT_Lib.DHC.Components.Plants.Cooling.CoolingPlant;
  model CoolingPlant
    "Isolated central cooling plant template model."
    extends Buildings.Experimental.DHC.Plants.Cooling.Examples.ElectricChillerParallel(
    redeclare Buildings.Fluid.Chillers.Data.ElectricEIR.ElectricEIRChiller_York_YT_1055kW_5_96COP_Vanes perChi(
      mEva_flow_nominal=mCHW_flow_nominal,
      mCon_flow_nominal=mCW_flow_nominal,
      QEva_flow_nominal=QChi_nominal),
    redeclare Buildings.Experimental.DHC.Plants.Cooling.ElectricChillerParallel pla,
    dTApp=3,
    mCHW_flow_nominal=18.3,
    dpCHW_nominal=44800,
    mMin_flow=0.03,
    mCW_flow_nominal=34.7,
    dpCW_nominal=46200,
    PFan_nominal=4999,
    TCHWSet=281.15,
    dpCooTowVal_nominal=5999,
    dpCHWPumVal_nominal=5999,
    dpCWPumVal_nominal=5999);
    annotation (
      Icon(
        coordinateSystem(
          preserveAspectRatio=false)),
      Diagram(
        coordinateSystem(
          preserveAspectRatio=false)),
      experiment(
        StopTime=86400,
        Tolerance=1e-06),
      Documentation(
        info="<html>
        <p>This model validates the district central cooling plant template model implemented in
  <a href=\"GMT_Lib.DHC.Components.Plants.Cooling.CoolingPlant.mo\">
  GMT_Lib.DHC.Components.Plants.Cooling.CoolingPlant.mot</a>.
  </p>
  </html>",
        revisions="<html>
  <ul>
  <li>
  October 20, 2021 by Chengnan Shi:<br/>
  First implementation.
  </li>
  </ul>
  </html>"),
      __Dymola_Commands(
        file="CoolingPlant.mos"
        "Simulate and Plot"));
  end CoolingPlant;
  '''
# ---
